design/src/rfcs/rfc_0020_endpoints_20.md

```diff
@@ -0,0 +1,599 @@
+be wired in properly.)
+
+### Converting a Smithy Endpoint to an AWS Endpoint
+
+A Smithy endpoint has an untyped, string->`Document` collection of properties. We need to interpret these properties to
+handle actually resolving an endpoint. We will implement this by implementing `TryFrom` from the Smithy endpoint to the
+AWS endpoint:
+> Note: It's possible `AWSEndpoint` will be removed entirely and the concept will be solely expressed in middleware. If
+> this is the case, this code would be ported into middleware directly.
+
+<details>
+<summary>Converting from smithy-native `Endpoint` to `AwsEndpoint`</summary>
+
+```rust
+impl TryFrom<Endpoint> for AwsEndpoint {
+    type Error = Box<dyn std::error::Error + Send + Sync>;
+
+    fn try_from(value: Endpoint) -> Result<Self, Self::Error> {
+        let uri: Uri = value.url().parse()?;
+        let base_endpoint = SmithyEndpoint::mutable(uri);
+        // look for v4 as an auth scheme
+        let auth_schemes = match value
+            .properties()
+            .get("authSchemes")
+            .ok_or("no auth schemes in metadata")?
+        {
+            Document::Array(schemes) => schemes,
+            _other => return Err("expected an array for authSchemes".into()),
+        };
+        let v4 = auth_schemes
+            .iter()
+            .flat_map(|doc| match doc {
+                Document::Object(map)
+                if map.get("name") == Some(&Document::String("v4".to_string())) =>
+                    {
+                        Some(map)
+                    }
+                _ => None,
+            })
+            .next()
+            .ok_or("could not find v4 as an acceptable auth scheme")?;
+
+        let signing_scope = match v4
+            .get("signingScope")
+            .ok_or_else(|| format!("signing scope missing: {:?}", v4))?
+        {
+            Document::String(s) => s.clone(),
+            _ => return Err("expected error".into()),
+        };
+        let signing_service = match v4.get("signingName").ok_or("signing service missing")? {
+            Document::String(s) => s.clone(),
+            _ => return Err("expected error".into()),
+        };
+        let credential_scope = CredentialScope::builder()
+            .region(Region::new(signing_scope))
+            .service(signing_service)
+            .build();
+        Ok(AwsEndpoint::new(base_endpoint, credential_scope))
+    }
+}
+```
+
+</details>
+
+### Implementing the rules engine
+
+<!-- Include a checklist of all the things that need to happen for this RFC's implementation to be considered complete -->
+Changes checklist
+-----------------
+**Rules Engine**
+
+- [ ] Endpoint rules code generator
+- [ ] Endpoint params code generator
+- [ ] Endpoint tests code generator
+- [ ] Implement ruleset standard library functions as inlineables. Note: pending future refactoring work, the `aws.`
```

Comment:
Are the standard rule functions going into inlineables to keep them hidden?

Reply:
yep